### ISSUE TITLE
Node/Result Filtering is duplicated

### ISSUE THREAD

**Issue by bryfox** [codaco/Network-Canvas#300]
We have two new interfaces, each of which involves filtering Nodes or results based on a user's text input. See #293 and #267.

Fuzzy search is required for #293, but is configurable as implemented, so it could be disabled for roster list to match current behavior.

**Comment by rebeccamadsen** [comment]
Does the fuzzy search allow filter to work across multiple fields? In the current implementation, the roster list is filtering based on the values in any displayed field on the cards.

**Comment by bryfox** [comment]
It does; the 'autocomplete' protocol specifies fields like this:

```
          "displayLabel": "nickname",
          "displayProperties": ["name", "age"],
```

... which appears to differ from the roster list protocol.

Perhaps we want to standardize on a format here? I like the one used in roster list (`nodeDisplay`...).

**Comment by jthrilly** [comment]
Either approach is fine! Sorry for the mismatch...

**Comment by rebeccamadsen** [comment]
I see that now. A standard would be great! I'm not picky on format. I know I personally debated whether to have a `sortable` true/false param for each display variable, or whether to have a `sortableFields` array elsewhere. I leaned towards the true/false param so there was less chance of typos, but the array has merits too.

**Comment by jthrilly** [comment]
Okay, here is how I propose we standardize these:

## List

``` JSON
{
  "id": "namegenroster2",
  "type": "NameGeneratorList",
  "label": "NG Classmates",
  "showExistingNodes": false,
  "prompts": [
    {
      "id": "7tk",
      "text": "Which classmates would you talk to outside of class?",
      "additionalAttributes": {
        "talk_friend": true
      },
      "subject": {
        "entity": "node",
        "type": "person"
      },
      "dataSource": "schoolPupils",
      "cardOptions": {
        "displayLabel": "nickname",
        "additionalProperties": ["name", "age"]
      },
      "sortOptions": {
        "sortOrder": {
          "name": "ASC"
        },
        "sortableProperties": [
          {
            "label": "Name",
            "variable": "name"
          },
          {
            "label": "Age",
            "variable": "age"
          }
        ]
      },
      "searchOptions": {
        "fuzzyness": 0.4,
        "matchProperties": ["nickname", "name"]
      }
    }
  ]
}
```

## Autocomplete

``` JSON
{
  "id": "namegenroster1",
  "type": "NameGeneratorAutoComplete",
  "label": "NG Classmates",
  "prompts": [
    {
      "id": "7cl",
      "text": "Which classmates would you also think of as friends?",
      "additionalAttributes": {
        "special_category": 46,
        "close_friend": true
      },
      "subject": {
        "entity": "node",
        "type": "person"
      },
      "dataSource": "schoolPupils",
      "cardOptions": {
        "displayLabel": "nickname",
        "additionalProperties": ["name", "age"]
      },
      "searchOptions": {
        "fuzzyness": 0.4,
        "matchProperties": ["nickname", "name"]
      }
    }
  ]
}
```

The key part is that both interfaces now have this:

``` JSON
      "searchOptions": {
        "fuzzyness": 0.4,
        "matchProperties": ["nickname", "name"]
      }
```

...which will give us a consistent API. The list interface can default `fuzziness` to 0.

## Other changes...

I know this is not ideal, but I took the opportunity to clean up the APIs somewhat, and make them more consistent. Please can you also look through the new configuration object supplied above, and make your interfaces compatible with this syntax?

The major thing that has changed is that I put the display configuration for the cards in a dedicated `cardOptions` property, so that again this can be consistent between the interfaces.

**Comment by bryfox** [comment]
One question/point on the new proposed format:

The Autocomplete interface displays `node` data as both `Node`s and `Card`s. I assume display label should continue to affect both displays, so `"cardOptions"` feels too specific.

Right now, list uses:

```
"nodeDisplay": {
  "primary": "nickname",
  "secondary": [
...
```

If that's not quite right, then maybe something in between...

```
"displayOptions": {
  "label": "nickname",
  "additionalProperties": [
...
```

**Comment by rebeccamadsen** [comment]
Roster list migrated to cardOptions with #303. There is a displayVariable in the variableRegistry that I nodes are intended to use (but are not yet using to my knowledge).

**Comment by bryfox** [comment]
Yeah, thank you. I was looking in an outdated branch. I stuck with `cardOptions` for now (in #267). I think requirements for node display might settle a bit with #305 and #311; we can update the protocol definition at that time if it makes sense.
